compiler/ballerina-lang/src/main/java/org/wso2/ballerinalang/compiler/semantics/analyzer/TypeChecker.java

```diff
@@ -1098,66 +1098,135 @@ private BType inferTableMemberType(List<BType> memTypes, BType expType) {
-            } else if (requiredFieldNames.contains(fieldName) && keySpecifierFieldNames.contains(fieldName)) {
-                field.symbol.flags = Flags.asMask(EnumSet.of(Flag.REQUIRED)) + Flags.asMask(EnumSet.of(Flag.READONLY));
-            } else if (requiredFieldNames.contains(fieldName)) {
-                field.symbol.flags = Flags.asMask(EnumSet.of(Flag.REQUIRED));
+                resultantField.symbol.flags = Flags.asMask(EnumSet.of(Flag.OPTIONAL));
+            } else if (keySpecifierFieldNames.contains(fieldName)) {
+                resultantField.symbol.flags = Flags.asMask(EnumSet.of(Flag.REQUIRED)) +
+                        Flags.asMask(EnumSet.of(Flag.READONLY));
+            } else {
+                resultantField.symbol.flags = Flags.asMask(EnumSet.of(Flag.REQUIRED));
+            }
+
+            inferredFields.add(resultantField);
+        });
+
+        return createTableConstraintRecordType(inferredFields, restFieldTypes, tableConstructorExpr.pos);
+    }
+
+    /**
+     * Gives unique {@code BType} list, extracted from {@code BField} list in which all fields having same field name.
+     *
+     * @param fields the list of same name fields
+     * @return a {@code List<BType>}
+     */
+    private List<BType> getUniqueTypesFromSameNameFieldList(List<BField> fields) {
+        List<BType> uniqueTypes = new ArrayList<>();
+
+        for (BField field : fields) {
+            BType type = field.getType();
+            if (type.tag == TypeTags.UNION) {
+                ((BUnionType) type).getMemberTypes().forEach(t -> {
+                    if (isUniqueType(uniqueTypes, t)) {
+                        uniqueTypes.add(t);
+                    }
+                });
+            } else if (isUniqueType(uniqueTypes, type)) {
+                uniqueTypes.add(type);
             }
         }
 
-        return createTableConstraintRecordType(allFieldSet, tableConstructorExpr.pos);
+        return uniqueTypes;
+    }
+
+    /**
+     * Replaces the {@code BField} type with a union type, derived from the given unique list of types.
+     *
+     * @param field       - bField of which, the type to be replaced
+     * @param uniqueTypes - list of unique types
+     * @return a {@code BField}
+     */
+    private BField replaceFieldType(BField field, List<BType> uniqueTypes) {
+        LinkedHashSet<BType> memberTypes = new LinkedHashSet<>(uniqueTypes);
+        BType unionType = BUnionType.create(null, memberTypes);
+
+        BVarSymbol originalSymbol = field.symbol;
+        BVarSymbol fieldSymbol = new BVarSymbol(originalSymbol.flags, originalSymbol.name, originalSymbol.pkgID,
+                unionType, originalSymbol.owner, originalSymbol.pos, VIRTUAL);
+
+        return new BField(field.name, field.pos, fieldSymbol);
+    }
+
+    private boolean hasOptionalFields(List<BField> fields) {
+        return fields.stream().anyMatch(field -> field.symbol.getFlags().contains(Flag.OPTIONAL));
```

Comment:
We try to avoid streams for trivial checks due to the potential impact on performance.